### PR TITLE
add action to get values from config.xml file on cordova projects

### PR DESCRIPTION
**Pull request by blackjid**
We've being using fastlane with our cordova projects and it works very good. 
We are missing some actions that could make our `Fastlane` files cleaner.

This is one of those actions.

With cordova you usually not edit the xcode project directly, instead use the `config.xml`
file.

### CODE REVIEW

**Comment by blackjid** [comment]
I could use some help here, @KrauseFx 

Where should I add this gem dependency??
over here?? https://github.com/fastlane/fastlane/blob/master/fastlane.gemspec#L34

**Comment by KrauseFx** [comment]
That's a great pull request, thanks @blackjid! 

> Where should I add this gem dependency??
over here?? https://github.com/fastlane/fastlane/blob/master/fastlane.gemspec#L34

no, you already added the `verify_gem!` call so no additional dependency is needed. Instead the user adds the dependency to their `Gemfile`

File: docs/Actions.md [hunk 1]
```diff
@@ -1492,6 +1493,24 @@ before_all do
...
+Get a value from a config file, which can be used to fetch the app identifier
+and more information about your app
+
+```ruby
+identifier = cordova_get_config_value(key: 'id')
+puts identifier # => tools.fastlane.cordova
+
+# path defaults to
+name = cordova_get_config_value(
+  path: 'cordova/config.xml',
```
Comment (KrauseFx) [inline docs/Actions.md]: Could you reformat that to look nicer, maybe in one line?

File: docs/Actions.md [hunk 2]
```diff
@@ -1492,6 +1493,24 @@ before_all do
...
+  key: 'name').first
+puts name.first # => Cordova App
```
Comment (KrauseFx) [inline docs/Actions.md]: why `.first`

File: docs/Actions.md [hunk 3]
```diff
@@ -1492,6 +1493,24 @@ before_all do
...
+and more information about your app
+
+```ruby
+identifier = cordova_get_config_value(key: 'id')
+puts identifier # => tools.fastlane.cordova
+
+# path defaults to
+name = cordova_get_config_value(
+  path: 'cordova/config.xml',
+  key: 'name').first
```
Comment (KrauseFx) [inline docs/Actions.md]: why `.first` here?

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+          config_file = File.open(File.expand_path(params[:path]))
+          config = XmlSimple.xml_in(config_file)
+
+          value = config[params[:key]]
+          Actions.lane_context[SharedValues::CORDOVA_GET_CONFIG_VALUE_CUSTOM_VALUE] = value
+
+          return value
+        rescue => ex
+          Helper.log.error ex
+          Helper.log.error "Unable to find config file at '#{path}'".red
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: That's not true. Also, please remove the `rescue` block as this should fail and raise an exception if something goes wrong.

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      #####################################################
+
+      def self.description
+        "Returns value from config.xml of your cordova project"
+      end
+
+      def self.details
+        # Optional:
+        # this is your change to provide a more detailed description of this action
+        "You can use this action to do cool things..."
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: Please remove `details`, if you don't want to provide details.

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      end
+
+      def self.available_options
+        # Define all options your action supports.
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: Please remove the 2 comments here

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                                       default_value: './config.xml',
+                                       verify_block: proc do |value|
+                                         path = File.expand_path(value)
+                                         raise "Couldn't find config file at path '#{value}'".red unless File.exist?(path)
+                                       end)
+        ]
+      end
+
+      def self.output
+        # Define the shared values you are going to provide
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: Please remove the comments here and also on the rest of this class

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        # you can do things like
+        #
+        #  true
+        #
+        #  platform == :ios
+        #
+        #  [:ios, :mac].include?(platform)
+        #
+
+        platform == :ios
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: Does this work on iOS only?

File: docs/Actions.md [hunk 9]
```diff
@@ -23,6 +23,7 @@ import './path/to/other/Fastfile'
 - [Using git](#using-git)
 - [Using mercurial](#using-mercurial)
 - [Notifications](#notifications)
+- [Cordova](#cordova)
```
Comment (KrauseFx) [inline docs/Actions.md]: This shouldn't be its own section, as it's only one action. Instead put it into Misc.

**Comment by blackjid** [comment]
Ok, I like that, but for the test?? is failing because the gem is not installed.

File: docs/Actions.md (hunk 9 above)
Comment (blackjid) [inline docs/Actions.md]: yes sorry, i'm thinking of build a few more actions for cordova project, and I rushed.
When that happens I will add the section again if it makes sense

Comment (KrauseFx) [inline docs/Actions.md]: Sounds good

**Comment by KrauseFx** [comment]
Don't worry about the Travis tess

File: docs/Actions.md (hunk 2 above)
Comment (blackjid) [inline docs/Actions.md]: ok, i'm using xmlsimple to parse the xml, so it return a simple ruby object that we can use in the `Fastfile`

The thing is that node inside a element are returned as an array.
In this case

```xml
<?xml version='1.0' encoding='utf-8'?>
<widget id="tools.fastlane.cordova" version="1.1.0" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>Cordova App</name>
    <description />
....
</widget>
```

The action is returning the a specific key (`key` parameter in the action). An in this example the real value is of `name` is an array, thats why I was using `.first`

```ruby
cordova_get_config_value(key: 'id') #  ["Cordova App"]
```

File: lib/fastlane/actions/cordova_get_config_value.rb (hunk 8 above)
Comment (blackjid) [inline lib/fastlane/actions/cordova_get_config_value.rb]: I've added `:android` also, probably those are the platforms on which this action makes sense

File: docs/Actions.md (hunk 3 above)
Comment (blackjid) [inline docs/Actions.md]: ok, i'm using xmlsimple to parse the xml, so it return a simple ruby object that we can use in the `Fastfile`

The thing is that node inside a element are returned as an array.
In this case

```xml
<?xml version='1.0' encoding='utf-8'?>
<widget id="tools.fastlane.cordova" version="1.1.0" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <name>Cordova App</name>
    <description />
....
</widget>
```

The action is returning the a specific key (`key` parameter in the action). An in this example the real value is of `name` is an array, thats why I was using `.first`

```ruby
cordova_get_config_value(key: 'id') #  ["Cordova App"]
```

File: docs/Actions.md (hunk 2 above)
Comment (blackjid) [inline docs/Actions.md]: it's duplicated, I've removed it

**Comment by blackjid** [comment]
What do you think??

This is the scenario, the most useful values to get from the `config.xml` file are probably `name` and `identifier`. 
Maybe `version`, but for that it makes more sense an action to update the version. 

Does it make any sense to you to have specific actions for that values??

```ruby
cordova_get_app_id # tools.fastlane.cordova
cordova_get_name # Cordova App
cordova_update_version(value: '1.2.1') 
```
Maybe they can alse set some `lane_context` that other actions can use, specially the `app_id`

Or with this generic action would be enough??

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      end
+
+      def self.output
+        [
+          ['CORDOVA_GET_CONFIG_VALUE_CUSTOM_VALUE', 'Value for the key required from the config.xml file']
+        ]
+      end
+
+      def self.return_value
+        # If you method provides a return value, you can describe here what it does
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: Maybe add one line saying, you get the result here, which might be a string or array (right?)

File: docs/Actions.md [hunk 11]
```diff
@@ -1736,3 +1736,16 @@ Install an Xcode plugin for the current user
...
+### cordova_get_config_value
+
+Get a value from a config file, which can be used to fetch the app identifier and more information about your app
+
+```ruby
+identifier = cordova_get_config_value(key: 'id')
+puts identifier # => tools.fastlane.cordova
+
+# path defaults to ./config.xml
+name = cordova_get_config_value(path: 'cordova/config.xml', key: 'name').first
```
Comment (KrauseFx) [inline docs/Actions.md]: Why is there a `.first` at the end of the line?

File: docs/Actions.md (hunk 3 above)
Comment (KrauseFx) [inline docs/Actions.md]: Ahh, okay :+1:

File: lib/fastlane/actions/cordova_get_config_value.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      end
+
+      def self.authors
+        ["platanus", "blackjid"]
+      end
+
+      def self.is_supported?(platform)
+        [:ios, :android].include?(platform)
```
Comment (KrauseFx) [inline lib/fastlane/actions/cordova_get_config_value.rb]: If it supports Mac as well, you can replace the whole line with just `true`

**Comment by KrauseFx** [comment]
Hey, thanks for updating the pull request. I added some final comments before it's ready to be merged. Could you also squash your commits so that we only add one commit to the history?

Thanks again for your great contribution :+1:

**Comment by ohwutup** [comment]
hey @blackjid, just wanted to check in here. would love to get this merged! would you mind just addressing @KrauseFx final comments and squash the commits and we will be good to go! :rocket:

**Comment by blackjid** [comment]
hey @ohwutup sorry about that... I'm not really sure if I like this approach for getting that from the config file... I find it complicated to generalize because you need to understand how the xml parser works and how it return its values...

for example

```
name = cordova_get_config_value(path: 'cordova/config.xml', key: 'name').first
```

Here I have to call first, because it returns an array of n tags with the name "name"....

I'm thinking on other approach, on having actions for specific keys. I'm doing this in https://github.com/platanus/fastlane-cordova/tree/initial-release/fastlane/actions
Instead of having a generic action `cordova_get_config_value` to have specific actions `cordova_get_app_name`, `cordova_get_app_ip`, `cordova_get_app_version`

I would love to know what do you think....

**Comment by ohwutup** [comment]
i don't have too much context because I'm not really familiar with cordova config files, but that seems like a much safer idea! would you want to get those merged in?

**Comment by blackjid** [comment]
cordova config files are just xml files...
what do you think on having both?? I could rebase this generic one, and then make a PR with the specific/safer ones?? 

cc: @KrauseFx ?

**Comment by ohwutup** [comment]
IMO, we should just go with the specific ones to avoid too many actions doing similar things.

**Comment by ohwutup** [comment]
IMO, we should just go with the specific ones to avoid too many actions doing similar things.

**Comment by ohwutup** [comment]
Hey @blackjid! Just wanted to bump this PR to see if we should close this in lieu of your upcoming PR with safer XML reading? :octopus:

**Comment by blackjid** [comment]
yes, you can close it...;)

**Comment by ohwutup** [comment]
Cool, thanks @blackjid! Looking forward to the other implementation! This is a great feature!

**Comment by jfbloom22** [comment]
+1 would love to see better cordova support.  thanks!

**Comment by blackjid** [comment]
I created a PR with some basic actions for cordova. #3754

**Pull request closed without merging**